File: user-analytics/appointment-service/src/main/java/com/healthcare/appointmentservice/service/impl/AppointmentService.java
/*
 * © 2025 Yevhen Tereshchenko
 * All rights reserved.
 *
 */

package com.healthcare.appointmentservice.service.impl;

import com.healthcare.appointmentservice.dto.AppointmentDTO;
import com.healthcare.appointmentservice.entity.AppointmentStatus;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;

public interface AppointmentService {
    AppointmentDTO createAppointment(AppointmentDTO appointmentDTO);

    @Transactional(readOnly = true)
    AppointmentDTO getAppointmentById(Long id);

    @Transactional(readOnly = true)
    Page<AppointmentDTO> getAllAppointments(Pageable pageable);

    @Transactional(readOnly = true)
    Page<AppointmentDTO> getAppointmentsByPatient(Long patientId, Pageable pageable);

    @Transactional(readOnly = true)
    Page<AppointmentDTO> getAppointmentsByDoctor(Long doctorId, Pageable pageable);

    @Transactional(readOnly = true)
    Page<AppointmentDTO> getAppointmentsByStatus(
            AppointmentStatus status, Pageable pageable);

    @Transactional(readOnly = true)
    List<AppointmentDTO> getAppointmentsBetween(
            LocalDateTime startDate, LocalDateTime endDate);

    @Transactional(readOnly = true)
    List<AppointmentDTO> getTodaysAppointments();

    AppointmentDTO updateAppointment(Long id, AppointmentDTO appointmentDTO);

    AppointmentDTO rescheduleAppointment(Long id, LocalDateTime newDateTime);

    AppointmentDTO cancelAppointment(Long id, String reason);

    AppointmentDTO confirmAppointment(Long id);

    AppointmentDTO completeAppointment(Long id, String notes);

    void deleteAppointment(Long id);
}

File: user-analytics/healthcare-frontend/src/app/services/appointment.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Appointment, AppointmentDTO, AppointmentStatus, PageResponse } from '../models/appointment.model';

@Injectable({
  providedIn: 'root'
})
export class AppointmentService {
  private readonly API_URL = 'http://localhost:8083/appointment-service/api/v1/appointments';
  private http = inject(HttpClient);

  private getHeaders(): HttpHeaders {
    const token = localStorage.getItem('token');
    return new HttpHeaders({
      'Authorization': `Bearer ${token}`
    });
  }

  createAppointment(appointment: AppointmentDTO): Observable<Appointment> {
    return this.http.post<Appointment>(this.API_URL, appointment, { headers: this.getHeaders() });
  }

  getAppointmentById(id: number): Observable<Appointment> {
    return this.http.get<Appointment>(`${this.API_URL}/${id}`, { headers: this.getHeaders() });
  }

  getAllAppointments(page: number = 0, size: number = 10): Observable<PageResponse<Appointment>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<PageResponse<Appointment>>(this.API_URL, { params, headers: this.getHeaders() });
  }

  getAppointmentsByPatient(patientId: number, page: number = 0, size: number = 10): Observable<PageResponse<Appointment>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<PageResponse<Appointment>>(`${this.API_URL}/patient/${patientId}`, { params, headers: this.getHeaders() });
  }

  getAppointmentsByDoctor(doctorId: number, page: number = 0, size: number = 10): Observable<PageResponse<Appointment>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<PageResponse<Appointment>>(`${this.API_URL}/doctor/${doctorId}`, { params, headers: this.getHeaders() });
  }

  getAppointmentsByStatus(status: AppointmentStatus, page: number = 0, size: number = 10): Observable<PageResponse<Appointment>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());
    return this.http.get<PageResponse<Appointment>>(`${this.API_URL}/status/${status}`, { params, headers: this.getHeaders() });
  }

  getAppointmentsBetween(startDate: string, endDate: string): Observable<Appointment[]> {
    const params = new HttpParams()
      .set('startDate', startDate)
      .set('endDate', endDate);
    return this.http.get<Appointment[]>(`${this.API_URL}/between`, { params, headers: this.getHeaders() });
  }

  getTodaysAppointments(): Observable<Appointment[]> {
    return this.http.get<Appointment[]>(`${this.API_URL}/today`, { headers: this.getHeaders() });
  }

  updateAppointment(id: number, appointment: AppointmentDTO): Observable<Appointment> {
    return this.http.put<Appointment>(`${this.API_URL}/${id}`, appointment, { headers: this.getHeaders() });
  }

  rescheduleAppointment(id: number, newDateTime: string): Observable<Appointment> {
    const params = new HttpParams().set('newDateTime', newDateTime);
    return this.http.patch<Appointment>(`${this.API_URL}/${id}/reschedule`, null, { params, headers: this.getHeaders() });
  }

  cancelAppointment(id: number, reason?: string): Observable<Appointment> {
    const params = reason ? new HttpParams().set('reason', reason) : new HttpParams();
    return this.http.patch<Appointment>(`${this.API_URL}/${id}/cancel`, null, { params, headers: this.getHeaders() });
  }

  confirmAppointment(id: number): Observable<Appointment> {
    return this.http.patch<Appointment>(`${this.API_URL}/${id}/confirm`, null, { headers: this.getHeaders() });
  }

  completeAppointment(id: number, notes?: string): Observable<Appointment> {
    const params = notes ? new HttpParams().set('notes', notes) : new HttpParams();
    return this.http.patch<Appointment>(`${this.API_URL}/${id}/complete`, null, { params, headers: this.getHeaders() });
  }

  deleteAppointment(id: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/${id}`, { headers: this.getHeaders() });
  }
}
